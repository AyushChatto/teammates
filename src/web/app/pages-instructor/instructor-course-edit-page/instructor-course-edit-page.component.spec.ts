import { HttpClientTestingModule } from '@angular/common/http/testing';
import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { AbstractControl, FormArray, FormBuilder, FormGroup, FormsModule } from '@angular/forms';
import { MatSnackBarModule } from '@angular/material';
import { RouterTestingModule } from '@angular/router/testing';
import { NgbModule } from '@ng-bootstrap/ng-bootstrap';
import { AjaxLoadingModule } from '../../components/ajax-loading/ajax-loading.module';
import { LoadingRetryModule } from '../../components/loading-retry/loading-retry.module';
import { LoadingSpinnerModule } from '../../components/loading-spinner/loading-spinner.module';
import { TeammatesCommonModule } from '../../components/teammates-common/teammates-common.module';
import { TeammatesRouterModule } from '../../components/teammates-router/teammates-router.module';
import {
  CustomPrivilegeSettingPanelComponent,
} from './custom-privilege-setting-panel/custom-privilege-setting-panel.component';
import { InstructorCourseEditPageComponent } from './instructor-course-edit-page.component';
import {
  InstructorEditPanelComponent,
} from './instructor-edit-panel/instructor-edit-panel.component';
import { ViewRolePrivilegesModalComponent } from './view-role-privileges-modal/view-role-privileges-modal.component';

describe('InstructorCourseEditPageComponent', () => {
  let component: InstructorCourseEditPageComponent;
  let fixture: ComponentFixture<InstructorCourseEditPageComponent>;
  const fb: FormBuilder = new FormBuilder();
  const instructorPrivileges: any = {
    coowner: {
      courseLevel: {
        canmodifycourse: true,
        canmodifyinstructor: true,
        canmodifysession: true,
        canmodifystudent: true,
        canviewstudentinsection: true,
        canviewsessioninsection: true,
        cansubmitsessioninsection: true,
        canmodifysessioncommentinsection: true,
      },
      sectionLevel: {
        firstSection: {
          canviewstudentinsection: true,
          canviewsessioninsection: true,
          cansubmitsessioninsection: true,
          canmodifysessioncommentinsection: true,
        },
        secondSection: {
          canviewstudentinsection: true,
          canviewsessioninsection: true,
          cansubmitsessioninsection: true,
          canmodifysessioncommentinsection: true,
        },
      },
      sessionLevel: {
        sessionOne: {
          canviewsessioninsection: true,
          cansubmitsessioninsection: true,
          canmodifysessioncommentinsection: true,
        },
        sessionTwo: {
          canviewsessioninsection: true,
          cansubmitsessioninsection: true,
          canmodifysessioncommentinsection: true,
        },
      },
    },
    manager: {
      courseLevel: {
        canmodifycourse: false,
        canmodifyinstructor: true,
        canmodifysession: true,
        canmodifystudent: true,
        canviewstudentinsection: true,
        canviewsessioninsection: true,
        cansubmitsessioninsection: true,
        canmodifysessioncommentinsection: true,
      },
      sectionLevel: {
        firstSection: {
          canviewstudentinsection: true,
          canviewsessioninsection: true,
          cansubmitsessioninsection: true,
          canmodifysessioncommentinsection: false,
        },
        secondSection: {
          canviewstudentinsection: true,
          canviewsessioninsection: true,
          cansubmitsessioninsection: false,
          canmodifysessioncommentinsection: true,
        },
      },
      sessionLevel: {
        sessionOne: {
          canviewsessioninsection: true,
          cansubmitsessioninsection: true,
          canmodifysessioncommentinsection: false,
        },
        sessionTwo: {
          canviewsessioninsection: true,
          cansubmitsessioninsection: false,
          canmodifysessioncommentinsection: true,
        },
      },
    },
    observer: {
      courseLevel: {
        canmodifycourse: false,
        canmodifyinstructor: false,
        canmodifysession: false,
        canmodifystudent: false,
        canviewstudentinsection: true,
        canviewsessioninsection: true,
        cansubmitsessioninsection: true,
        canmodifysessioncommentinsection: true,
      },
      sectionLevel: {
        firstSection: {
          canviewstudentinsection: true,
          canviewsessioninsection: false,
          cansubmitsessioninsection: true,
          canmodifysessioncommentinsection: true,
        },
        secondSection: {
          canviewstudentinsection: false,
          canviewsessioninsection: true,
          cansubmitsessioninsection: true,
          canmodifysessioncommentinsection: true,
        },
      },
      sessionLevel: {
        sessionOne: {
          canviewsessioninsection: false,
          cansubmitsessioninsection: true,
          canmodifysessioncommentinsection: false,
        },
        sessionTwo: {
          canviewsessioninsection: false,
          cansubmitsessioninsection: false,
          canmodifysessioncommentinsection: false,
        },
      },
    },
    tutor: {
      courseLevel: {
        canmodifycourse: false,
        canmodifyinstructor: false,
        canmodifysession: false,
        canmodifystudent: false,
        canviewstudentinsection: false,
        canviewsessioninsection: false,
        cansubmitsessioninsection: false,
        canmodifysessioncommentinsection: false,
      },
      sectionLevel: {
        firstSection: {
          canviewstudentinsection: false,
          canviewsessioninsection: false,
          cansubmitsessioninsection: false,
          canmodifysessioncommentinsection: false,
        },
        secondSection: {
          canviewstudentinsection: false,
          canviewsessioninsection: false,
          cansubmitsessioninsection: false,
          canmodifysessioncommentinsection: false,
        },
      },
      sessionLevel: {
        sessionOne: {
          canviewsessioninsection: false,
          cansubmitsessioninsection: false,
          canmodifysessioncommentinsection: false,
        },
        sessionTwo: {
          canviewsessioninsection: false,
          cansubmitsessioninsection: false,
          canmodifysessioncommentinsection: false,
        },
      },
    },
    custom: {
      courseLevel: {
        canmodifycourse: false,
        canmodifyinstructor: true,
        canmodifysession: false,
        canmodifystudent: true,
        canviewstudentinsection: false,
        canviewsessioninsection: true,
        cansubmitsessioninsection: false,
        canmodifysessioncommentinsection: true,
      },
      sectionLevel: {
        firstSection: {
          canviewstudentinsection: true,
          canviewsessioninsection: false,
          cansubmitsessioninsection: true,
          canmodifysessioncommentinsection: false,
        },
        secondSection: {
          canviewstudentinsection: false,
          canviewsessioninsection: true,
          cansubmitsessioninsection: false,
          canmodifysessioncommentinsection: true,
        },
      },
      sessionLevel: {
        sessionOne: {
          canviewsessioninsection: false,
          cansubmitsessioninsection: true,
          canmodifysessioncommentinsection: false,
        },
        sessionTwo: {
          canviewsessioninsection: true,
          cansubmitsessioninsection: false,
          canmodifysessioncommentinsection: true,
        },
      },
    },
  };
  const instructorList: any[] = [
    {
      googleId: 'coowner@gmail.com',
      name: 'Ms. Co-Owner',
      email: 'coowner@email.com',
      role: 'Co-owner',
      isDisplayedToStudents: true,
      displayedName: 'Ayush The Co-Owner',
      privileges: instructorPrivileges.coowner,
    },
    {
      googleId: 'manager@gmail.com',
      name: 'Mr. Manager',
      email: 'manager@email.com',
      role: 'Manager',
      isDisplayedToStudents: true,
      displayedName: 'Byush the Manager',
      privileges: instructorPrivileges.manager,
    },
    {
      googleId: 'observer@gmail.com',
      name: 'Ms. Observer',
      email: 'observer@email.com',
      role: 'Observer',
      isDisplayedToStudents: false,
      displayedName: 'Cyush the Invisible Observer',
      privileges: instructorPrivileges.observer,
    },
    {
      name: 'Mr. Tutor',
      email: 'tutor@email.com',
      role: 'Tutor',
      isDisplayedToStudents: true,
      displayedName: 'Dyush the Tutor without Google ID',
      privileges: instructorPrivileges.tutor,
    },
    {
      googleId: 'custom@gmail.com',
      name: 'Ms. Custom',
      email: 'custom@email.com',
      role: 'Custom',
      isDisplayedToStudents: true,
      displayedName: 'Eyush the Custom',
      privileges: instructorPrivileges.custom,
    },
  ];

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [
        InstructorCourseEditPageComponent,
        InstructorEditPanelComponent,
        ViewRolePrivilegesModalComponent,
        CustomPrivilegeSettingPanelComponent,
      ],
      imports: [
        NgbModule,
        FormsModule,
        AjaxLoadingModule,
        TeammatesCommonModule,
        RouterTestingModule,
        HttpClientTestingModule,
        LoadingSpinnerModule,
        LoadingRetryModule,
        TeammatesRouterModule,
      ],
    })
    .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(InstructorCourseEditPageComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should snap with default fields', () => {
    expect(fixture).toMatchSnapshot();
  });

  it('should snap with populated fields', () => {
    component.timezone = 'GMT+8';
    component.timezones = ['GMT+8', 'GMT+7', 'GMT+6'];
    component.sectionNames = ['firstSection', 'secondSection'];
    component.feedbackNames = ['sessionOne', 'sessionTwo'];
    component.instructorList = instructorList;
    component.courseToEdit = {
      id: 'CS3281',
      name: 'Thematic Systems',
      timeZone: 'GMT+8',
    };
    component.formAddInstructor = fb.group({
      googleId: [''],
      name: [''],
      email: [''],
      isDisplayedToStudents: [{ value: true }],
      displayedName: ['InstructorNameAdd'],
      role: ['Co-owner'],
      privileges: instructorPrivileges.coowner,
      tunePermissions: fb.group({
        permissionsForCourse: fb.group({
          canmodifycourse: true,
          canmodifyinstructor: true,
          canmodifysession: true,
          canmodifystudent: true,
          canviewstudentinsection: true,
          canviewsessioninsection: true,
          cansubmitsessioninsection: true,
          canmodifysessioncommentinsection: true,
        }),
        tuneSectionGroupPermissions: fb.array([]),
      }),
    });
    component.formEditInstructors = fb.group({
      formInstructors: fb.array([]),
    });
    component.instructor = instructorList[0];
    component.formEditCourse = fb.group({
      id: [{ value: component.courseToEdit.id, disabled: true }],
      name: [{ value: component.courseToEdit.name, disabled: true }],
      timeZone: [{ value: component.courseToEdit.timeZone, disabled: true }],
    });

    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

});
