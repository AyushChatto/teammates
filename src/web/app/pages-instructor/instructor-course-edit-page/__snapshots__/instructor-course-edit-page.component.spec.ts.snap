// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InstructorCourseEditPageComponent should snap with default fields 1`] = `
<tm-instructor-course-edit-page
  courseService={[Function CourseService]}
  fb={[Function FormBuilder]}
  feedbackNames={[Function Array]}
  httpRequestService={[Function HttpRequestService]}
  instructorList={[Function Array]}
  instructorToShowIndex={[Function Number]}
  isAddingInstructor="false"
  isEditingCourse="false"
  navigationService={[Function NavigationService]}
  ngbModal={[Function NgbModal]}
  route={[Function ActivatedRoute]}
  router={[Function Router]}
  sectionNames={[Function Array]}
  statusMessageService={[Function StatusMessageService]}
  timezone={[Function String]}
  timezoneService={[Function TimezoneService]}
  timezones={[Function Array]}
  user="undefined"
>
  
</tm-instructor-course-edit-page>
`;

exports[`InstructorCourseEditPageComponent should snap with populated fields 1`] = `
<tm-instructor-course-edit-page
  courseService={[Function CourseService]}
  courseToEdit={[Function Object]}
  fb={[Function FormBuilder]}
  feedbackNames={[Function Array]}
  formAddInstructor={[Function FormGroup]}
  formEditCourse={[Function FormGroup]}
  formEditInstructors={[Function FormGroup]}
  httpRequestService={[Function HttpRequestService]}
  instructor={[Function Object]}
  instructorList={[Function Array]}
  instructorToShowIndex={[Function Number]}
  isAddingInstructor="false"
  isEditingCourse="false"
  navigationService={[Function NavigationService]}
  ngbModal={[Function NgbModal]}
  route={[Function ActivatedRoute]}
  router={[Function Router]}
  sectionNames={[Function Array]}
  statusMessageService={[Function StatusMessageService]}
  timezone={[Function String]}
  timezoneService={[Function TimezoneService]}
  timezones={[Function Array]}
  user="undefined"
>
  <div
    class="card bg-light top-padded"
  >
    <div
      class="card-header bg-primary text-white"
    >
      <strong>
        Course:
      </strong>
      <span
        class="float-sm-right"
      >
        
        <button
          class="btn btn-light btn-sm btn-left btn-course"
          ng-reflect-ngb-tooltip="Edit course name"
          ngbtooltip="Edit course name"
        >
          <i
            aria-hidden="true"
            class="fa fa-pen"
          />
           Edit 
        </button>
        
        <button
          class="btn btn-light btn-sm btn-course"
          ng-reflect-ngb-tooltip="Delete the course and its corr"
          ngbtooltip="Delete the course and its corresponding students and sessions"
        >
          <i
            aria-hidden="true"
            class="fa fa-trash"
          />
           Delete 
        </button>
      </span>
    </div>
    <div
      class="card-body fill-plain"
    >
      <form
        class="form form-horizontal ng-untouched ng-pristine"
        ng-reflect-form="[object Object]"
        novalidate=""
      >
        <div>
          <div
            class="form-row form-group"
          >
            <label
              class="col-sm-3"
            >
              Course ID:
            </label>
            <div
              class="col-sm-3"
            >
              <input
                class="col-xs-12 form-control ng-untouched ng-pristine"
                disabled=""
                formcontrolname="id"
                maxlength="40"
                ng-reflect-maxlength="40"
                ng-reflect-name="id"
                tabindex="1"
              />
            </div>
          </div>
          <div
            class="form-row form-group"
          >
            <label
              class="col-sm-3"
            >
              Course Name:
            </label>
            <div
              class="col-sm-9"
            >
              <input
                class="form-control ng-untouched ng-pristine"
                disabled=""
                formcontrolname="name"
                maxlength="64"
                ng-reflect-maxlength="64"
                ng-reflect-name="name"
                ng-reflect-ngb-tooltip="The name of the course, e.g. S"
                ngbtooltip="The name of the course, e.g. Software Engineering."
                tabindex="2"
              />
            </div>
          </div>
          <div
            class="form-row form-group"
          >
            <label
              class="col-xs-12 col-sm-3"
            >
              Time Zone:
            </label>
            <div
              class="col-xs-12 col-sm-9"
            >
              <div
                class="input-group"
              >
                <select
                  class="form-control ng-untouched ng-pristine"
                  disabled=""
                  formcontrolname="timeZone"
                  ng-reflect-name="timeZone"
                  ng-reflect-ngb-tooltip="The time zone for the course. "
                  ngbtooltip="The time zone for the course. You should not need to change this as it is auto-detected based on your
                  device settings. TEAMMATES automatically adjusts to match the current time offset in your area, including clock changes due to daylight saving time."
                >
                  
                  <option
                    ng-reflect-value="GMT+8"
                    value="GMT+8"
                  >
                    GMT+8
                  </option>
                  <option
                    ng-reflect-value="GMT+7"
                    value="GMT+7"
                  >
                    GMT+7
                  </option>
                  <option
                    ng-reflect-value="GMT+6"
                    value="GMT+6"
                  >
                    GMT+6
                  </option>
                </select>
                <span
                  class="input-group-btn"
                >
                  <button
                    class="btn btn-primary"
                    disabled=""
                    type="button"
                  >
                     Auto-Detect 
                  </button>
                </span>
              </div>
            </div>
          </div>
          <div
            class="form-group"
          >
            <div
              class=" col-sm-12 align-center"
            >
              
            </div>
          </div>
        </div>
      </form>
    </div>
  </div><div
    class="ng-untouched ng-pristine ng-valid"
    ng-reflect-form="[object Object]"
  >
    <div
      class="ng-untouched ng-pristine ng-valid"
      formarrayname="formInstructors"
      ng-reflect-name="formInstructors"
    >
      
    </div>
  </div><div
    class="top-padded bottom-padded align-center"
  >
    
    <button
      class="btn btn-primary btn-instr"
    >
       Add New Instructor 
    </button>
  </div>
</tm-instructor-course-edit-page>
`;
